### ISSUE TITLE
Add 'None of these' feature to checkboxes

### ISSUE THREAD

**Issue by joelanman** [alphagov/govuk-design-system-backlog#188]
## What

Add guidance and possibly code to checkboxes to allow a 'None of these'

## Why

We previously had this guidance in Dropbox Paper. The alternative is the user selects no checkboxes and clicks continue, which is an implicit answer. This is arguably less clear than an explicit answer, and there is a risk the use clicked continue accidentally.

**Comment by alexnewmannn** [comment]
We need to implement this pattern on 2 pages in our service and will need to research how people deal with (we are agent facing). Our current idea is `None of these` will disable all other checkboxes and deselecting it will enable the checkboxes retaining their previous values (so that if you click `None` by accident you don't need to reenter other options). It might be that this is strange behaviour so will be testing it against `None of these` just deselecting everything else hopefully! Will keep this issue updated when we come to research it (hopefully in the upcoming weeks)

**Comment by edwardhorsford** [comment]
My service has an example of this. Screenshot below.

<img width="518" alt="screenshot 2019-01-30 at 10 51 58" src="https://user-images.githubusercontent.com/2204224/51976647-1eeef700-247d-11e9-8da6-2fe0f60ef47f.png">

If you check none, all other options are cleared. If you check one of the other options, none is cleared. This requires javascript so we'll also have to check this server side.

---

NB - more generically you might call this a 'mutually exclusive' option. Any checkbox that is set to mutually exclusive will automatically clear others when set (and vice versa)

**Comment by terrysimpson99** [comment]
This would work the same for 'All'.
![Screenshot_20190411-095818_Opera](https://user-images.githubusercontent.com/48089049/55944628-eb7cd580-5c40-11e9-9bf6-27e83b4ac0c8.jpg)

**Comment by peteryates** [comment]
I struggle to see what having a "None of the above" option _achieves_. 'None of the above' and leaving all the checkboxes blank _mean the same thing_, and that is what causes confusion.

Isn't having a one checkbox that doesn't act like any of the other checkboxes an antipattern? Perhaps that's too strong a word, but it's at least a case of material dishonesty.

**Comment by alexnewmannn** [comment]
@peteryates Yeh you're right, they do mean the same thing. The main difference, at least for us, is that one is explicit and one isn't. How can you tell the difference between accidentally submitting without checking an answer and actively not answering that question? You could obviously go down the route of "how do you know they selected any of them correctly?" But this feels like a good middle ground of removing possible user error.

**Comment by edwardhorsford** [comment]
@peteryates this allows the user to positively give an answer that's clearly for them. Without it, it may be unclear for them how to proceed.

I actually needed this today on a government form - none of the options seemed appropriate - but without an option _like_ this, I may have picked one thinking that I needed to pick the most appropriate.

**Comment by fofr** [comment]
@edwardhorsford In your design, have you tried separating the last item from the others with "or" text, like the Radio example: https://design-system.service.gov.uk/components/radios/#radio-items-with-a-text-divider

I wonder if this text helps separate the different kind of option from the list, or if users would misinterpreting checkboxes as behaving like radios because of that text.

**Comment by spattisson** [comment]
There is also the option for the checkboxes to be preceded with a qualifying question? So before @edwardhorsford 's example something like "Do you know about some parts of the supply chain?" - that being a Yes or No question would allow the effective "none of the above" answer to be given.
Downside would be the addition of more questions to a journey if many instances of checkbox questions were necessary.

**Comment by joelanman** [comment]
I think what might help define when you use a guard question (Do you X?) or a 'None of the above' would be how common the situation would be for users. So if it's common to have None, use a guard question, but if it's rare, use a None of the above, instead of making all users answer a guard question.

**Comment by peteryates** [comment]
Using a guard question seems like the cleanest approach to me.

The idea of having one check box that acts differently to the others, more like a radio button, just doesn't sit right.

**Comment by LucaDelBuonoHMCTS** [comment]
I wrote a little ES6 script to deal with this:
```
(function() {
    'use strict';

    const $selectCheckboxes = document.querySelectorAll('[data-mutually-exclusive="js-select-choice"]');
    const $deselectCheckboxes = document.querySelectorAll('[data-mutually-exclusive="js-deselect-choice"]');

    for (const selectedElement of $selectCheckboxes) {
        selectedElement.addEventListener('click', () => {
            for (const elementToDeselect of $deselectCheckboxes) {
                elementToDeselect.checked = false;
            }
        }, false);
    }

    for (const selectedElement of $deselectCheckboxes) {
        selectedElement.addEventListener('click', () => {
            for (const elementToDeselect of $selectCheckboxes) {
                elementToDeselect.checked = false;
            }
        }, false);
    }
}).call(this);
```

**Comment by edwardhorsford** [comment]
Here's an example from one of the new Covid services (GDS made?):
<img width="873" alt="Screenshot 2020-04-03 at 11 08 11" src="https://user-images.githubusercontent.com/2204224/78349506-9d708880-759b-11ea-8f4a-b038fba198ad.png">

**Comment by joelanman** [comment]
Yes this is from Offer Coronavirus support from your business - I think its a good example of where None is useful. It's not easy to ask a yes/no question here as the question would be 'Can you offer expertise' when we're interested in specific types.

**Comment by LucaDelBuonoHMCTS** [comment]
This is what I needed it for in the upcoming Equality questionnaire:

![pcq](https://user-images.githubusercontent.com/42244677/78368196-26e38300-75bb-11ea-9ca3-02a94ffd4c04.png)

**Comment by edwardhorsford** [comment]
Sent via @ wayneddgu - [another example](https://register-trailer.service.gov.uk/eligibility/travelling_abroad) of a service with an `or` option in checkboxes.

<img width="1151" alt="Screenshot 2020-04-15 at 11 17 58" src="https://user-images.githubusercontent.com/2204224/79326612-0a284300-7f0b-11ea-82e3-df502e7d052b.png">

**Comment by joelanman** [comment]
To expand more on when I think this pattern is necessary - its when there's no obvious guard question for the list. So 'Do you have a car?' easily guards 'What kind of car do you have?'. But there's no obvious guard question for 'These conditions' or 'These countries' as in the examples above.

**Comment by timpaul** [comment]
It would be great if the people on this thread could let us know what logic they applied for users who don't have JavaScript enabled.
